src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -30,52 +42,105 @@ def execute(self, reqStatus):
-                    requestStatuses[wname] = transferRecords
-            self.updateTransferInfo(requestStatuses)
+                        "+++ request %s transfers are completed", reqName)
+            self.updateTransferInfo(transferRecords, completedWorkflows)
+            # finally, update statuses for requests
+            for reqName, status in reqStatus.items():
+                self.change(reqName, status, '+++ monit')
         except Exception as err:  # general error
             self.logger.exception('+++ monit error: %s', str(err))
 
-    def getTransferRecords(self, reqName):
+    def completion(self, transferRecords, pileup=None):
+        """
+        Helper function to calculate completion of given records
+        :param records: list of status records
+        :param pileup: check pileup completion (optional)
+        :return: completion status
+        """
+        # check completion of all transfers
+        status = False
+        transferTypes = ['primary', 'secondary']
+        for rec in transferRecords:
+            campaign = rec['campaignName']
+            cdict = self.getCampaignConfig(campaign)
+            thr = cdict.get('PartialCopy', 0)
+            if pileup and rec['dataType'] not in transferTypes:
+                continue
+            if rec['completion'] >= thr:
+                status = True
+        return status
+
+    def getCampaignConfig(self, campaign):
+        "Helper function to get given campaign from backend or cache"
+        if campaign not in self.campaigns:
+            campaigns = self.reqmgrAux.getCampaignConfig("ALL_DOCS")
+            for cdict in campaigns:
+                if cdict not in self.campaigns:
+                    self.campaigns[cdict['CampaignName']] = cdict
+        return self.campaigns.get(campaign, {})
+
+    def updateTransferInfo(self, transferRecords, completedWorkflows):
+        """
+        Update transfer records in backend
+        :param transferRecords: list of transfer records
+        :param completedWorkflows: list of completed workflow names to be removed from backend
         """
-        Get transfer records for given request name from CouchDB.
-        Transfer records on backend has the following form
-        https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
-        So far logic is based on option D of the records
+        tstamp = time.time()
+        for doc in transferRecords:
+            if doc['workflowName'] in completedWorkflows:
+                continue  # we skip completed workflows
+            transfers = []
+            for rec in doc.get('transfers', []):
+                # obtain new transfer ids for given request name and dataset
+                tids, completion = self.getTransferIds(rec['dataset'])
+                rec.update({'transferIDs': tids, 'completion': completion})
+                transfers.append(rec)
+            wname = doc['workflowName']
+            doc['timestamp'] = tstamp
```

Comment:
We don't use such key in the transfer doc.

Reply:
replaced with lastupdate